engine/hooks/gerrit.go

```diff
@@ -0,0 +1,274 @@
+		if gerritEvent.PatchSet.Author != nil {
+			if gerritEvent.PatchSet.Author.Username != "" {
+				payload["git.author"] = gerritEvent.PatchSet.Author.Username
+			}
+			if gerritEvent.PatchSet.Author.Email != "" {
+				payload["git.author.email"] = gerritEvent.PatchSet.Author.Email
+			}
+		}
+	}
+	// change-merged
+	if gerritEvent.NewRev != "" {
+		payload["git.hash"] = gerritEvent.NewRev
+	}
+	payload["gerrit.type"] = gerritEvent.Type
+
+	payload["payload"] = string(e.GerritEvent.Message)
+
+	d := dump.NewDefaultEncoder(&bytes.Buffer{})
+	d.ExtraFields.Type = false
+	d.ExtraFields.Len = false
+	d.ExtraFields.DetailedMap = false
+	d.ExtraFields.DetailedStruct = false
+	d.Formatters = []dump.KeyFormatterFunc{dump.WithDefaultLowerCaseFormatter()}
+	payloadValues, errDump := d.ToStringMap(payload)
+	if errDump != nil {
+		return nil, sdk.WrapError(errDump, "Cannot dump payload %+v ", payload)
+	}
+	h.Payload = payloadValues
+
+	return &h, nil
+}
+
+func (s *Service) ComputeGerritStreamEvent(ctx context.Context, vcsServer string, gerritEventChan <-chan GerritEvent) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case e := <-gerritEventChan:
+			var repo string
+			switch {
+			case e.Change != nil:
+				repo = e.Change.Project
+			case e.RefUpdate != nil:
+				repo = e.RefUpdate.Project
+			}
+
+			hooks, err := s.Dao.FindGerritTasksByRepo(vcsServer, repo)
+			if err != nil {
+				log.Error("ComputeGerritStreamEvent > Unable to list task for repo %s/%s", vcsServer, repo)
+				continue
+			}
+
+			msg, err := json.Marshal(e)
+			if err != nil {
+				log.Error("unable to marshal gerrit event: %v", err)
+			}
+
+			for _, h := range hooks {
+				if !sdk.IsInArray(e.Type, h.Events) {
+					continue
+				}
+
+				//Load the task
+				gerritHook := s.Dao.FindTask(h.UUID)
+				if gerritHook == nil {
+					log.Error("Unknown uuid %s", h.UUID)
+					continue
+				}
+
+				//Prepare a web hook execution
+				exec := &sdk.TaskExecution{
+					Timestamp: time.Now().UnixNano(),
+					UUID:      h.UUID,
+					GerritEvent: &sdk.GerritEventExecution{
+						Message: msg,
+					},
+				}
+
+				//Save the web hook execution
+				s.Dao.SaveTaskExecution(exec)
+
+				//Push the webhook execution in the queue, so it will be executed
+				s.Dao.EnqueueTaskExecution(exec)
+			}
+		}
+	}
+}
+
+// ListenGerritStreamEvent listen the gerrit event stream
+func ListenGerritStreamEvent(ctx context.Context, v sdk.VCSConfiguration, gerritEventChan chan<- GerritEvent) {
+	signer, err := ssh.ParsePrivateKey([]byte(v.Password))
+	if err != nil {
+		log.Error("unable to read ssh key: %v", err)
+		return
+	}
+
+	// Create config
+	config := &ssh.ClientConfig{
+		User: v.Username,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(signer),
+		},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+
+	url, _ := url2.Parse(v.URL)
+
+	// Dial TCP
+	conn, err := ssh.Dial("tcp", fmt.Sprintf("%s:%d", url.Hostname(), v.SSHPort), config)
```

Comment:
does this kind of connection work through a ssh bastion ?

Reply:
no need for the moment